gtda/metaestimators/collection_transformer.py

```diff
@@ -0,0 +1,140 @@
+"""CollectionTransformer meta-estimator."""
+# License: GNU AGPLv3
+
+from functools import reduce
+from operator import and_
+
+import numpy as np
+from joblib import Parallel, delayed
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.base import clone
+from sklearn.utils.metaestimators import if_delegate_has_method
+
+from gtda.utils import check_collection
+
+
+class CollectionTransformer(BaseEstimator, TransformerMixin):
+    """Meta-transformer for applying a fit-transformer to each input in a
+    collection.
+
+    If `transformer` possesses a ``fit_transform`` method,
+    ``CollectionTransformer(transformer)`` also possesses a
+    :meth:`fit_transform` method which, on each entry in its input ``X``,
+    fit-transforms a clone of `transformer`. A collection (list or ndarray) of
+    outputs is returned.
+
+    Parameters
+    ----------
+    transformer : object
```

Comment:
Interesting. This was somewhat deliberate as I was thinking about users who build their own transformer and forget to inherit from `TransformerMixin`. Because in principle we do not even the transformer to have a `fit` or a `transform` method separately, just a `fit_transform`, which rather makes it pointless to inherit from `TransformerMixin`, no?

Reply:
Right now, there are no checks being conducted on the transformer. We need at least to check the existence of `fit_transform` as below:

https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/pipeline.py#L877

I would also argue that the transformer should have `set_params` and `get_params` (or rather inherit from `BaseEstimator`), just to ensure that you can do grid searches.

Reply:
Agree with checking the existence of fit_transform, thanks for the suggestion.

Concerning the second point, the class already inherits from `BaseEstimator`.

Reply:
I don't mean the class, but the transformer itself.

Reply:
Oh I see. In these matters, I prefer to be more lenient. Even scikit-learn's `Pipeline` and `FeatureUnion` don't check for this. The user might not be interested in grid searches and still know what they're doing. And even without the transformer being a BaseEstimator one can do grid-searches provided one does not try to go deep inside the transformer parameters with the `__` syntax.

What do you think about a warning?

Reply:
`Pipeline` and `FeatureUnion` derive from `BaseComposition` which handle a bit the `__` logic. I don't think our user would understand what's going on if they try to do a grid search but can't iterate over hyperparameters.

A warning would be fine, but if they use GridSearchCV, it's likely that it won't even be printed.

Reply:
I've added a validation step which raises an error if `fit_transform` does not exist, and a warning if the transformer is not a `BaseEstimator` instance.